### DISCUSSION TITLE
Terragrunt: Matching base_source_url in .hcl files

### DISCUSSION THREAD

**Discussion by syphernl** [renovatebot/renovate#13401] [Request Help]
I'm having some problems getting Renovate to pick up some "special" Terragrunt files.

## Terragrunt Structure

Our Terragrunt structure (simplified) looks like:

```
├── _envcommon
│   └── vpc.hcl
├── staging
│   ├── eu-west-1
│   │   └── vpc
│   │       └── terragrunt.hcl
```

In `_envcommon/vpc.hcl` we have a definition of the module and the version to use:

```hcl
locals {
  base_source_url = "git::git@gitlab.com:ORG/REPO/terraform-aws-components.git//modules/vpc?ref=v0.2.0"
}
```

Within the `staging/region/vpc/terragrunt.hcl` we refer to this value:

```
include "envcommon" {
  path   = "${dirname(find_in_parent_folders())}/_envcommon/vpc.hcl"
  expose = true
}

terraform {
  source = "${include.envcommon.locals.base_source_url}"
}
```

## Renovate configuration

However, as the default Terragrunt manager is looking for a source but gets an unresolved variable it doesn't treat this as an update.

The regex I have added to the Renovate regexManager:

```js
    {
      description: 'Handles additional Terragrunt files',
      fileMatch: ["^(_envcommon\|shared)\/[^/]+\.hcl$"],
      matchStrings: [
        'base_source_url\s*?=\s*?"(?<lookupName>.*)\/\/(?<depName>.*)\?ref=(?<currentValue>.*)"',
      ],
      versioningTemplate: 'semver',
    },
```

The file is picked up by the regexManager:

```
DEBUG: Matched 1 file(s) for manager regex: _envcommon/vpc.hcl (repository=ORG/REPO/infra)
```

But no matches are found and no versions are being updated either.

In an online regex tester I have  resolves to the following:

| Match | Value |
| ------------- | ------------- |
| lookupName | `git::git@gitlab.com:ORG/REPO/terraform-aws-components.git` |
| depName | `modules/vpc` |
| currentValue | `v0.2.0` |

Could it be that Renovate does not expect the `lookupName` to be an URL but expects something like `ORG/REPO/terraform-aws-components`?
I have also tried this `matchString` to no avail:

```js
'base_source_url\s*?=\s*?"git::git@gitlab.com:(?<lookupName>.*).git\/\/modules\/(?<depName>.*)\?ref=(?<currentValue>.*)"'
```

This one produces:

| Match | Value |
| ------------- | ------------- |
| lookupName | `ORG/REPO/terraform-aws-components` |
| depName | `vpc` |
| currentValue | `v0.2.0` |

**Comment by viceice** [comment]
I think you need to put the full url into depName

**Reply by viceice** [reply]
You also need to assign a datasource

**Reply by syphernl** [reply]
Do you mean switching around the values of `lookupName` and `depName`? Or simply getting rid of `lookupName` and only using `depName`?

Regarding datasource: I'm now using 

```
datasourceTemplate: 'terraform-module',
```



----


```js
    {
      description: 'Handles additional Terragrunt files',
      fileMatch: ["^(_envcommon\|shared)\/[^/]+\.hcl$"],
      matchStrings: [
        'base_source_url\s*?=\s*?"(?<depName>.*)\/\/(?<moduleName>.*)\?ref=(?<currentValue>.*)"',
      ],
      versioningTemplate: 'semver',
      datasourceTemplate: 'terraform-module',
    },
```

**Reply by viceice** [reply]
Yes only use depName, the datasource should be able to split the url

**Reply by syphernl** [reply]
This is what I have now:

```js
    {
      description: 'Handles additional Terragrunt files',
      fileMatch: ["^(_envcommon\|shared)\/[^/]+\.hcl$"],
      matchStrings: [
        'base_source_url\s*?=\s*?"(?<depName>.*)\?ref=(?<currentValue>.*)"',
      ],
      versioningTemplate: 'semver',
      datasourceTemplate: 'terraform-module',
    },
```

The regex produces:

| Match | Value |
|--------|------|
| `depName` | `git::git@gitlab.com:ORG/REPO/terraform-aws-components.git//modules/vpc` |
| `currentVersion` | `v0.2.0` |

Unfortunately still nothing gets picked up.

I've also tried `terragrunt`  as `datasourceTemplate` but no luck with that either.

**Reply by viceice** [reply]
Terragrunt is no datasource, terraform-module datasource probably don't support your url.

Can you test by not using a variable and normal terragrunt manager?

**Reply by syphernl** [reply]
I think I probably shouldn't be using the `terraform-module` datasource but the `gitlab-releases` or `gitlab-tags` instead.. 

No luck with that either unfortunatly..

```js
    {
      description: 'Handles additional Terragrunt files',
      fileMatch: ["^(_envcommon\|shared)\/[^/]+\.hcl$"],
      matchStrings: [
        'base_source_url\s*?=\s*?"git::git@gitlab.com:(?<lookupName>.*).git\/\/(?<friendlyName>.*)\?ref=(?<currentValue>.*)"',
      ],
      versioningTemplate: 'semver',
      datasourceTemplate: 'gitlab-tags',
    },
```

It's quite a bit of trial & error, is there a way to debug the regexManager? I've already set `LOG_LEVEL=debug` but it doesn't provide much insight as to what it finds (or why it doesn't).

**Reply by phuongdnguyen** [reply]
@syphernl  Did you resolve this one?
